graphbolt/include/graphbolt/rowwise_pick.h

```diff
@@ -0,0 +1,184 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file graphbolt/rowwise_pick.h
+ * @brief Contains the function definition for RowWisePickPerEtype.
+ */
+
+#ifndef GRAPHBOLT_ROWWISE_PICK_H_
+#define GRAPHBOLT_ROWWISE_PICK_H_
+
+#include <graphbolt/csc_sampling_graph.h>
+#include <graphbolt/utils.h>
+
+using namespace graphbolt::utils;
+
+namespace graphbolt {
+namespace sampling {
+
+using CSCPtr = CSCSamplingGraph*;
+using RangePickFn = std::function<torch::Tensor(
+    int64_t start, int64_t end, int64_t num_samples)>;
+using TensorList = std::vector<torch::Tensor>;
+static constexpr int kDefaultPickGrainSize = 2;
+
+/**
+ * @brief Performs row-wise picking based on the given parameters.
+ *
+ * @param graph The pointer to a csc sampling graph.
+ * @param rows The tensor containing the row indices.
+ * @param num_picks The tensor containing the number of picks per edge type.
+ * @param probs Optional tensor containing probabilities for picking.
+ * @param return_eids Boolean indicating if edge IDs need to be returned. The
+ * last TensorList in the tuple is this value when required.
+ * @param replace Boolean indicating if picking is done with replacement.
+ * @param pick_fn The function used for picking.
+ * @return A tuple containing the picked rows, picked columns, picked etypes,
+ * and picked edge IDs (if required).
+ */
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor>
+RowWisePickPerEtype(
+    const CSCPtr graph, const torch::Tensor& rows,
+    const torch::Tensor& num_picks, const torch::optional<torch::Tensor>& probs,
+    bool return_eids, bool replace, RangePickFn pick_fn) {
+  const auto indptr = graph->CSCIndptr();
+  const auto indices = graph->Indices();
+  const auto type_per_edge = graph->TypePerEdge().value();
+  const int64_t num_rows = rows.size(0);
+  const int64_t num_etypes = num_picks.size(0);
+  std::vector<int64_t> picked_num_per_row(num_rows + 1);
+  TensorList picked_cols_per_row(
+      num_rows, torch::tensor({}, indices.options()));
+  TensorList picked_etypes_per_row(
+      num_rows, torch::tensor({}, type_per_edge.options()));
+  TensorList picked_eids_per_row;
+  if (return_eids) {
+    picked_eids_per_row.resize(num_rows, torch::tensor({}, indptr.options()));
+  }
+  int64_t min_num_pick = -1;
+  auto num_picks_accessor = num_picks.accessor<int64_t, 1>();
```

Comment:
Is is possible that `num_picks` uses a dtype other than int64_t?

Reply:
changed to column.
And this could be any bits int, but this transforamtion is safe and the scalar value will not take much more memory.